CONFIGURATION.md
# Options

## layout

Hangul layout name

### default

`dubeolsik`

## esc_turn_off

Turn off hangul mode when esc is pressed especially for VIM users

### default

`true`

## hangul_keys

Keycodes for switch hangul mode

### default

`[Hangul, Henkan, AltR]`

## xim_preedit_font

Preedit window font name and size for XIM

### default

`[D2Coding, 15.0]`

## gtk_commit_english

Commit english rather then bypass keyevent need for some programs like gedit

### default

`true`

## compose

Adjust compose, decompose jamo

### compose_choseong_ssang

```
ㄱ + ㄱ = ㄲ
ㅅ + ㅅ = ㅆ
ㄷ + ㄷ = ㄸ
ㅂ + ㅂ = ㅃ
ㅈ + ㅈ = ㅉ
```

#### default

`true`

### decompose_choseong_ssang

Same as above but work on backspace(e.g. ㄲ -> ㄱ)

#### default

`false`

### compose_jungseong_ssang

```
ㅑ + ㅣ = ㅒ
ㅕ + ㅣ = ㅖ
```

#### default

`false`


### decompose_jungseong_ssang

#### default

`false`

### compose_jongseong_ssang

```
ㄱ + ㄱ = ㄲ
ㅅ + ㅅ = ㅆ
```

#### default

`false`

### decompose_jongseong_ssang

#### default

`false`


docs/CONFIGURATION.md
# Options

## layout

Hangul layout name

### default

`dubeolsik`

## esc_turn_off

Turn off hangul mode when esc is pressed especially for VIM users

### default

`true`

## hangul_keys

Keycodes for switch hangul mode

### default

`[Hangul, Henkan, AltR]`

## xim_preedit_font

Preedit window font name and size for XIM

### default

`[D2Coding, 15.0]`

## compose

Adjust compose, decompose jamo

### compose_choseong_ssang

```
ㄱ + ㄱ = ㄲ
ㅅ + ㅅ = ㅆ
ㄷ + ㄷ = ㄸ
ㅂ + ㅂ = ㅃ
ㅈ + ㅈ = ㅉ
```

#### default

`true`

### decompose_choseong_ssang

Same as above but work on backspace(e.g. ㄲ -> ㄱ)

#### default

`false`

### compose_jungseong_ssang

```
ㅑ + ㅣ = ㅒ
ㅕ + ㅣ = ㅖ
```

#### default

`false`


### decompose_jungseong_ssang

#### default

`false`

### compose_jongseong_ssang

```
ㄱ + ㄱ = ㄲ
ㅅ + ㅅ = ㅆ
```

#### default

`false`

### decompose_jongseong_ssang

#### default

`false`


CHANGELOG.md
# Dev

* Rework build scripts
* Fix NumLock bug again
* Add GTK2, GTK4 immodule

# 0.8.1

* Fix GTK3 space commit bug
* Fix Qt5 backspace bug
* Use C in GTK3

# 0.8.0

* Add Qt5 immodule
* Complete moum backspace (e.g. ㅚ -> ㅗ, ㅞ -> ㅜ)

# 0.7.0

* Fix hangul bug on NumLock, CapLock, ScrollLock
* Fix xim preedit window delete bug
* Fix deb file
* Add `--log` option in `kime-xim`
* Add size in `xim-preedit-font`

# 0.6.0

* Don't reset on XIM set_ic_values
* Redraw when XIM preedit state changed
* Update packaging scripts
* Add Install guide
* Now kime-xim print version when pass `--version`

# 0.5.1

* Fix XIM modifier bug

# 0.5.0

* Now `dubeolsik`, `sebeolsik-390`, `sebeolsik-391` layouts are embeded and no need local file
* Fix intellij issue
* Fix Super key bug
* Package deb file 

# 0.4.1

* Fix unhandled keycode doesn't occur reset 

# 0.4.0

* Create CHANGELOG
* Decrease binary sizes with make engine cdylib
* Reset when focus_out (XIM)
* Add compose config
* Fix XIM start bug when XIM_SERVER is not set
* Support Control modifiers
* Support Super modifiers
* Commit forwarded events

# 0.3.0

* Add `compose_ssangjaum` config
* Allow `Hangul` key
* Reset when focus_out (GTK3)
* Fix firefox backspace bug
* Fix firefox enter, esc bug

# 0.2.1

* Bypass shift keys

# 0.2.0

* Support compose jungseong (ㅑ + ㅣ = ㅒ)
* Fix jongseong to next choseong when jungseong is entered (옹 + ㅏ = 오아)
* Support ESC to disable hangul mode for VIM users
* Bypass ctrl chars
* Implement config file
* Reset when unhandled keysym is entered
* Add sebeolsik 390, 391 layouts


engine/core/src/config.rs
use crate::{keycode::Key, KeyCode, Layout};
use ahash::AHashSet;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(default)]
pub struct ComposeConfig {
    pub compose_choseong_ssang: bool,
    pub decompose_choseong_ssang: bool,
    pub compose_jungseong_ssang: bool,
    pub decompose_jungseong_ssang: bool,
    pub compose_jongseong_ssang: bool,
    pub decompose_jongseong_ssang: bool,
}

impl Default for ComposeConfig {
    fn default() -> Self {
        Self {
            compose_choseong_ssang: true,
            decompose_choseong_ssang: false,
            compose_jungseong_ssang: false,
            decompose_jungseong_ssang: false,
            compose_jongseong_ssang: false,
            decompose_jongseong_ssang: false,
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Serialize, Deserialize)]
#[repr(C)]
pub enum ModuleType {
    Xim,
    Wayland,
    Gtk,
    Qt,
    Unknown,
}

impl Default for ModuleType {
    fn default() -> Self { Self::Unknown }
}

#[derive(Serialize, Deserialize)]
#[serde(default)]
pub struct RawConfig {
    layout: String,
    esc_turn_off: bool,
    hangul_keys: Vec<String>,
    xim_preedit_font: (String, f64),
    notify_modules: AHashSet<ModuleType>,
    compose: ComposeConfig,
}

impl Default for RawConfig {
    fn default() -> Self {
        const DEFAULT_HANGUL_KEYS: &[Key] = &[
            Key::normal(KeyCode::AltR),
            Key::normal(KeyCode::Muhenkan),
            Key::normal(KeyCode::Hangul),
            Key::super_(KeyCode::Space),
        ];

        Self {
            layout: "dubeolsik".to_string(),
            esc_turn_off: true,
            hangul_keys: DEFAULT_HANGUL_KEYS
                .iter()
                .map(ToString::to_string)
                .collect(),
            notify_modules: [ModuleType::Xim, ModuleType::Wayland]
                .iter()
                .copied()
                .collect(),
            xim_preedit_font: ("D2Coding".to_string(), 15.0),
            compose: ComposeConfig::default(),
        }
    }
}

pub struct Config {
    pub(crate) layout: Layout,
    pub(crate) esc_turn_off: bool,
    pub(crate) hangul_keys: AHashSet<Key>,
    pub(crate) compose: ComposeConfig,
    pub(crate) notify_modules: AHashSet<ModuleType>,
    xim_preedit_font: (String, f64),
}

impl Default for Config {
    fn default() -> Self {
        Self::new(Layout::default(), RawConfig::default())
    }
}

impl Config {
    pub fn new(layout: Layout, raw: RawConfig) -> Self {
        Self {
            layout,
            esc_turn_off: raw.esc_turn_off,
            compose: raw.compose,
            hangul_keys: raw
                .hangul_keys
                .iter()
                .filter_map(|s| s.parse().ok())
                .collect(),
            notify_modules: raw.notify_modules,
            xim_preedit_font: raw.xim_preedit_font,
        }
    }

    pub fn xim_preedit_font(&self) -> &(String, f64) {
        &self.xim_preedit_font
    }

    pub fn from_raw_config(raw: RawConfig, dir: Option<xdg::BaseDirectories>) -> Self {
        let layout = dir
            .and_then(|dir| {
                dir.list_config_files("layouts")
                    .into_iter()
                    .find_map(|layout| {
                        if layout.file_stem()?.to_str()? == raw.layout {
                            Some(Layout::from_items(
                                serde_yaml::from_reader(std::fs::File::open(layout).ok()?).ok()?,
                            ))
                        } else {
                            None
                        }
                    })
            })
            .unwrap_or_else(|| {
                // User layout not exists fallback to embeded layouts
                match raw.layout.as_str() {
                    "dubeolsik" => Layout::load_from(include_str!("../data/dubeolsik.yaml"))
                        .expect("Load dubeolsik layout"),
                    "sebeolsik-390" => {
                        Layout::load_from(include_str!("../data/sebeolsik-390.yaml"))
                            .expect("Load sebeolsik-390 layout")
                    }
                    "sebeolsik-391" => {
                        Layout::load_from(include_str!("../data/sebeolsik-391.yaml"))
                            .expect("Load sebeolsik-391 layout")
                    }
                    // custom layout
                    other => {
                        eprintln!("Can't find layout {}", other);
                        Layout::default()
                    }
                }
            });

        Self::new(layout, raw)
    }

    pub fn load_from_config_dir() -> Option<Self> {
        let dir = xdg::BaseDirectories::with_prefix("kime").ok()?;

        let raw = dir
            .find_config_file("config.yaml")
            .and_then(|config| serde_yaml::from_reader(std::fs::File::open(config).ok()?).ok())
            .unwrap_or_default();

        Some(Self::from_raw_config(raw, Some(dir)))
    }
}
